### ISSUE TITLE
kube-proxy:  too many iptables input filter chain will cause performence issue

### ISSUE THREAD

**Issue by xujieasd** [kubernetes/kubernetes#56842]
**Is this a BUG REPORT or FEATURE REQUEST?**:

 /kind bug

**What happened**:

As we know, for these serivces, nodeports, externalIPs without endpoint, kube-proxy iptables will make rejection rules in filter tables, it can avoid `close wait issue` #43212.
all these filter rules will match `KUBE-SERVICES` chain in both `INPUT` and `OUTPUT` chain
(`"-A INPUT -j KUBE-SERVICES"`, `"-A OUTPUT -j KUBE-SERVICES"`)

However, we find if there are large numbers of filter rules, will badly impact network performence.
We do some tests and find the impact mainly cause by INPUT chain. After manually remove rule "`INPUT -j KUBE-SERVICES`", the network performence will have a greate improvement

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

we have an enviroment with more than 1000 reject rules,
use iperf test pod2pod network bandwidth（Theoretical upper limit 1000Mb/s）
the bandwidth is only `50Mb/s` !
and after manually remove rule "`INPUT -j KUBE-SERVICES`", bandwidth result is about `600Mb/s`

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release): CentOS-7
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

/sig network
/area kube-proxy

**Comment by xujieasd** [comment]
After I have searched kube-proxy simily issues, I find there already have a PR, #56164
This PR splitting out the "new" part of KUBE-SERVICES (NodePort and ExternalIP reject rules) into a separate KUBE-EXTERNAL-SERVICES chain run from INPUT
However I think it still cannot avoid large numbers of reject rules in INPUT chain, if we have many services with NodePort and ExternalIP without enpoints

I think I have a better idea: 
1. use a new flag to mark those package which service have not endpoint in PREROUTING chain , just like what `KUBE-MARK-MASQ` do
2. in INPUT chain and OUTPUT chain, reject these packages with the special mark

e.g
1. in nat table, create rule like 
`-N KUBE-MARK-REJECT`
`-A KUBE-MARK-REJECT -j MARK --set-xmark 0x2000/0x2000`

2. in nat table, for service without endpoint, make rule like
`-A KUBE-SVC-3GHDLKRD75JCF45N -m comment --comment "default/stress-pod:clientport has no endpoints" -j KUBE-MARK-REJECT`

3. in filter table, keep `KUBE-SERVICES` in both INPUT and OUTPUT chain 
`-A INPUT -m comment --comment "kubernetes service portals" -j KUBE-SERVICES`
`-A OUTPUT -m comment --comment "kubernetes service portals" -j KUBE-SERVICES`
and reject package with the 0x2000/0x2000 mark
`-A KUBE-SERVICES -m comment --comment "kubernetes service traffic requiring reject" -m mark --mark 0x2000/0x2000 -j REJECT`

I've tried above method in our enviroment, and it works
I've already have a commit with this feature: https://github.com/kubernetes/kubernetes/compare/master...xujieasd:KUBE-MARK-REJECT
If it does make sense, I'll submit a RP

**Comment by m1093782566** [comment]
please xref: https://github.com/kubernetes/kubernetes/pull/56164

**Comment by m1093782566** [comment]
@thockin @danwinship 

wdyt?

**Comment by m1093782566** [comment]
BTW, @xujieasd have you ever tried IPVS mode? I believe it won't create any REJECT rule.

**Comment by xujieasd** [comment]
@m1093782566,
I didn't ever try, seems IPVS mode doesn't create any REJECT rule in filter table.
However, I think it's worth to have a try to make sure that IPVS mode could avoid `close_wait` issue

**Comment by m1093782566** [comment]
> However, I think it's worth to have a try to make sure that IPVS mode could avoid close_wait issue

I can confirm IPVS mode has no such issue.  I have tested it.

**Comment by danwinship** [comment]
It would be nice if we could avoid using another mark bit (although also it doesn't make sense to do the address matching twice). I'm at kubecon this week but I'll look into this next week.

**Comment by xujieasd** [comment]
Do you mean check `KUBE-MARK-REJECT` mark bit in both INPUT and OUTPUT chain twice?

I think It doesn't matter, because the the KEY is time complexity has already reduced from `O(N) `to `O(1)`. Before add this mark bit, package have to scan the whole filter rules created by kube-proxy, but now we remove these filter rules, package only need to check itself.
I think one more additional matching in INPUT chain or OUTPUT chain will not impact networking performance, but can make the implementation simple

@danwinship 

> although also it doesn't make sense to do the address matching twice

**Comment by danwinship** [comment]
@xujieasd no, I was saying, I'd like to avoid using the extra mark bit, but on the other hand, using the extra mark bit actually makes a lot of sense, because right now we have one rule per IP in PREROUTING and then one rule per IP again in INPUT ("doing the address matching twice"), but using the mark would let us avoid that.

**Comment by danwinship** [comment]
@xujieasd does the "proxier-fixes-all" branch of https://github.com/danwinship/kubernetes/ work for you as well as your KUBE-MARK-REJECT branch does? It has 3 fixes:
- `KUBE-SERVICES` and `KUBE-EXTERNAL-SERVICES` are split, as in #56164
- Match rules for services with no endpoints are no longer added to the `nat` table (meaning now for each IP, a rule is added *either* to `nat` *or* to `filter`, but not both, so we don't compare against any IP multiple times).
- The `KUBE-SERVICES` and `KUBE-EXTERNAL-SERVICES` chains in `-t filter` now only get run on the first packet of new connections rather than all packets (since if that first packet doesn't get rejected any future packets aren't going to either).

I *think* this should have the same effect as your branch.

**Comment by xujieasd** [comment]
@danwinship 
emm...
I've checked your branch, if my understanding is correct, your branch:
1. use `-m conntrack --ctstate NEW` when create `KUBE-SERVICES` & `KUBE-EXTERNAL-SERVICES` chain in `filter table`
2. for each service without endpoints, still keep corresponding rules in `filter table`, but will not create any rules in `nat table` 

It sounds reasonable and can also avoid filter chain performence issue because filter rule will only get run on the first packet

**Comment by danwinship** [comment]
@xujieasd do you have an environment where you can actually test it and see if it's as good as your earlier branch? We've done some testing and found improvements in some cases but I don't know for sure that we're really reproducing your environment...

**Comment by xujieasd** [comment]
I'd like to help to test this, but our earlier environment is now used for doing some production work. These machines temporarily have no bandwidth to doing other things... hope you can understand
@danwinship
